### ISSUE TITLE
generate() output changes with batch composition (batch_size>1 not reproducible per-sample); also ~24x slower on H100 than A6000

### ISSUE THREAD

**Issue by g8a9** [MELT-proj/training#118]
## Summary

`MELTForCausalLM.generate()` produces **different completions for the same
sample depending on what batch it's generated in**, under fully greedy,
deterministic decoding (`do_sample=False`, no beam search). Comparing
batch_size=1 vs batch_size=16 for the same 64 samples, same checkpoint, same
prompts: **8 of 64 samples (12.5%) diverge**, and several of those are not
minor numerical noise (a flipped word here or there) but materially
different transcriptions. This means any batched generation run with this
model is not reliably measuring the model's actual per-sample output, only
something close to it in aggregate.

This was found while setting up an ASR/ST eval campaign in `melt-eval`
(MELT-proj/eval) and is being filed here because the likely cause is in this
repo's modeling code, not the eval harness.

## Reproduction

Checkpoint: `IFT-700both-w2vbF-llama1bInsT-mlpF-s1337-8g` (final checkpoint,
i.e. the run root, not a `checkpoint-N` subfolder) at
`/mnt/scratch-nyx/giuseppe/melt/outputs/IFT-700both-w2vbF-llama1bInsT-mlpF-s1337-8g`.

Two `inspect eval` runs, same 64-sample slice (`librispeech-clean` test,
first 64 samples of the frozen set at
`/mnt/scratch-artemis/giuseppe/melt-data/eval-sets/asr-eval-campaign-v1`),
same checkpoint, same prompts, differing only in `-M batch_size`:

```bash
# from the melt-eval checkout, venv at /mnt/scratch-artemis/giuseppe/venvs/melteval
inspect eval melteval/tasks.py@speech --model melt/<checkpoint> \
  --log-dir <out> \
  -T frozen_set=/mnt/scratch-artemis/giuseppe/melt-data/eval-sets/asr-eval-campaign-v1 \
  -T task_filter=asr -T dataset_id=librispeech-clean -T limit=64 \
  -M batch_size=1     # run A
  -M batch_size=16    # run B
```

Decoding is the harness's default for both runs: greedy (`do_sample=False`,
since no temperature is set), `max_new_tokens=256`, `use_cache=True`, no
beam search (`num_beams` is never passed, so it's HF's default of 1). See
`melteval/providers/melt.py::_generate_kwargs()` if useful context, but
nothing there differs between the two runs except `batch_size` itself
(which controls how many samples `MELTAPI._process_batches` groups into one
`model.generate()` call).

Full logs (`.eval`, viewable with `inspect view`):
- batch=1: `/mnt/scratch-artemis/giuseppe/melt-data/eval-logs-batch-consistency-check/2026-09-08T07-49-09-00-00_speech-asr-librispeech-clean_mgWPFk8xwTAoaEz6x4hnfs.eval`
- batch=16: `/mnt/scratch-artemis/giuseppe/melt-data/eval-logs-batch-consistency-check/2026-09-08T07-50-05-00-00_speech-asr-librispeech-clean_a7HtzrYv5EVgEHq3saFZCs.eval`

Aggregate corpus WER/CER over the 64 samples was nearly identical between
the two runs (bs=1: WER 0.227, CER 0.123; bs=16: WER 0.226, CER 0.121) --
**do not take that as evidence the two runs agree**. It's not: the
per-sample completions genuinely differ for 8/64 samples, and the aggregate
metrics landing close appears to be individual errors roughly cancelling
out in this particular 64-sample slice, not the model producing consistent
output. This should not be assumed to hold at a different sample count or
for a different checkpoint.

## The actual diffs

Two of the eight are not explainable as minor floating-point noise -- they
change the transcribed content:

```
sample 08-000001:
  bs=1 : "concord returns place emmitt's the tent."
  bs=16: "concord returns place 3rd with 2,000."

sample 08-000027:
  bs=1 : "her husband, the beaumont saint-françois, was she evidently dear about it."
  bs=16: "her husband, the beaumont same frame two, but she evidently didn't care about it."
```

The other six are smaller (single-word substitutions), which *could* be
ordinary floating-point non-associativity between single-sample and batched
matmul/attention kernels, but given the two above, I'd treat all eight as
suspect rather than assume most are benign:

```
sample 08-000004:  "white buckingham" -> "while buckingham"
sample 08-000015:  "this moment of his friend" -> "this movement of his friend"
                    "indescribe recus" -> "indescribe recless"
sample 08-000020:  "maniacal camp" -> "menicamp"
sample 08-000033:  "old clothes" -> "all clothes"; "regular house" -> "ragula house"
sample 08-000037:  "young kit much" -> "young kid much"; "in the as delight" -> "in the us deloitte"
sample 08-000048:  "border up house mystery happened" -> "border appears mystery, happen"
                    "april 6," -> "april 6th,"
```

56 of 64 samples were byte-identical between the two runs, for what it's
worth -- this isn't affecting every sample, just a meaningful minority.

## Where to look

This is very likely related to (possibly the same root cause as) a second,
already-investigated issue found the same day: `MELTForCausalLM.generate()`
is dramatically slower on H100 than on A6000 for the same checkpoint and
input, with no batching involved at all (single sample, batch_size=1):

| GPU | time for 20 generated tokens | notes |
|---|---|---|
| H100 (partition `h100`, node `dionysus`) | 27.00s (~1.35s/token) | batch_size=8 and 16 didn't even finish in 60s with the real prompt |
| A6000 (partition `a6000`) | 1.13s (~0.057s/token) | ~24x faster, same checkpoint/prompt/code |

Ruled out as the cause: device placement (GPU correctly allocated and used
on both), NFS/data loading (the A6000 node's checkpoint *load* was actually
10x *slower*, 92s vs 9.4s -- yet its `generate()` call was 24x *faster*,
so it isn't storage latency), and batch size in isolation (the H100 slowdown
reproduces even at batch=1). Reproduction scripts from that investigation
are still on artemis at `/mnt/scratch-artemis/giuseppe/profile_generate.py`
(synthetic short prompt) and `/mnt/scratch-artemis/giuseppe/profile_generate2.py`
(the real chat-templated prompt via `melteval.prompt`, which is what
actually triggers the H100 pathology -- the synthetic prompt is fast on
both GPUs). Both issues -- the H100 slowdown and this batch-dependent
output divergence -- point at the same code path:

- `melt/modeling/modeling_melt.py`, `MELTForCausalLM.generate()` (~line 1239):
  delegates to `self.text_decoder.generate(inputs_embeds=..., attention_mask=...)`
  after building `inputs_embeds`/`attention_mask` via `_get_audio_embeddings`
  + `_merge_embeddings`.
- `_merge_embeddings` -> `_inject_tensor` (~line 1052 onward): splices audio
  embeddings into the text embedding sequence at the `<|audio|>` placeholder
  position(s), per-batch-item, via a Python-level
  `for batch_idx in range(batch_size):` loop (~line 1416). Since
  audio-injected sequences within one batch have differing lengths (the
  audio placeholder expands to a variable number of audio-embedding
  positions depending on that sample's audio duration), the *left-padding*
  and *attention mask* this loop constructs per batch item is the natural
  place for a batch-composition-dependent bug to live: if the mask or
  position handling for one sample is subtly influenced by the *other*
  samples present in the same batch (their lengths, their padding amount,
  or a shared/incorrectly-broadcast tensor), that would produce exactly this
  symptom -- correct-ish output that changes depending on batch neighbors.
- Worth checking specifically: whether attention masks are correctly
  isolating each sample's own (audio + text) tokens from every other
  sample's padding in a batch; whether position_ids are computed
  per-sample accounting for that sample's own padding rather than a shared
  scheme; and whether the text decoder's cache/attention implementation
  (`_attn_implementation`, unset in this checkpoint's `config.json` so it
  resolves to whatever `transformers`' default is) handles a merged
  embedding sequence with left-padding + a variable-position audio
  injection correctly in a batched call.

## Impact

Any eval or inference run using `batch_size > 1` with this model produces
per-sample outputs that are not reproducible/trustworthy in isolation --
only some approximate aggregate may be usable, and even that isn't
guaranteed to be unbiased. This affects every downstream WER/CER/BLEU/etc.
number computed with batching enabled, which is the normal way to run
anything at scale on a GPU.

**Comment by g8a9** [comment]
Investigated. It is not a padding leak, and there are two separate causes. All
numbers below are on the checkpoint named in the issue, MN5 `acc_debug`, one
H100 64 GB, `melt_cuda126.sif` (transformers 5.16.1, torch 2.9.1+cu126),
bfloat16, greedy, librispeech clean/test from `shar-indexed`.

## The suggested cause is ruled out

Checked stage by stage rather than by inspection:

| stage | test | result |
|---|---|---|
| processor featurisation | per-sample `input_features` alone vs in a batch of 16 | bit-identical, 16/16 |
| audio encoder | audio right-padded to the batch's 1584 frames, still batch size 1 | transcripts byte-identical, 16/16 |
| encoder, merge, mask | the whole stack in float64 on CPU, lengths 37/120/501/1503 | max abs diff 2.6e-16 |
| `generate()` position ids | instrumented the decoder call | already per-row, derived from the merged mask, every row starts at 0 |
| chunking at `max_audio_seq_len` | 1499 and 1501 frames, alone vs padded to 3000 | identical, no NaN |
| merged-sequence left padding | batch size 1, 0 to 1024 masked positions prepended | changes 2 of 12 transcripts, not monotone in the padding |

So `_inject_tensor`, the encoder and the mask are all fine. Attention masks do
isolate each sample, and position ids are already computed per sample.

## Cause 1: the decode loop is not reproducible

The same call, same process, same inputs, five times:

| batch | frames | encoder | prefill forward | generate |
|---|---|---|---|---|
| 1 | 72 | 1 | 1 | 1 |
| 2 | 128 | 1 | 1 | 1 |
| 4 | 172 | 1 | 1 | 1 |
| 8 | 308 | 1 | 1 | 1 |
| 16 | 1584 | 1 | 1 | **5 of 5 distinct** |

Distinct bit patterns, so 1 means reproducible. Rows 6, 8, 12 and 13 moved, and
rows 6 and 8 are exactly two of the rows that "differed between batch 1 and
batch 16". A good part of the 8/64 in the issue is the batched run not
repeating itself, rather than a systematic batch effect.

The audio encoder is bit-stable and so is a single full-sequence decoder forward
over identical merged embeddings. Only the autoregressive loop varies. Pinning
the backend at batch 16 identifies it:

| configuration | distinct of 5 | s/run |
|---|---|---|
| sdpa, backend chosen by torch (what ran) | 4 | 1.416 |
| sdpa pinned to CUDNN_ATTENTION | 3 | 1.415 |
| sdpa pinned to EFFICIENT_ATTENTION | 1 | 1.611 |
| sdpa pinned to MATH | 1 | 3.881 |
| `attn_implementation='eager'` | 1 | 2.351 |
| `attn_implementation='flash_attention_2'` | 1 | **1.333** |

cuDNN attention is the non-deterministic one and torch's default picks it.
`torch.use_deterministic_algorithms(True, warn_only=True)` with
`CUBLAS_WORKSPACE_CONFIG=:4096:8` does not fix it and emits no warning.

The trigger is a **mixed-length** batch, not batch size. Replicating one
utterance so every row is the same length is reproducible at batch 16 even for
a 1630-token sequence:

| utterance | frames | batch 1 | 2 | 4 | 8 | 16 |
|---|---|---|---|---|---|---|
| 1.42 s | 72 | 1 | 1 | 1 | 1 | 1 |
| 6.87 s | 344 | 1 | 1 | 1 | 1 | 1 |
| 31.65 s | 1584 | 1 | 1 | 1 | 1 | 1 |

Mixed lengths make the merged mask non-trivial, flash attention refuses a
non-null mask, and SDPA falls through to cuDNN.

**And nothing could ask for a different implementation.**
`MELTPreTrainedModel` declared `_supports_flash_attn = False`, so
`from_pretrained(attn_implementation="flash_attention_2")` was refused outright,
and `attn_implementation=` otherwise landed on the composite config and never
reached the backbones, so `"eager"` silently produced two sdpa sub-models.
Training escaped this only because `train.py` re-applies the value onto the
decoder sub-config by hand (#86). The `False` dates from a rename commit
(10887a1) and predates flash attention in this codebase.

## Cause 2: arrival-order batching

Exactly 32 decode steps in every cell, so only the batch varies.

| batch | uniform s/sample | arrival order s/sample | length-bucketed s/sample | padding share |
|---|---|---|---|---|
| 1 | 0.2423 | 0.2448 | 0.2445 | 1% |
| 8 | 0.0337 | 0.0573 | 0.0370 | 58% |
| 16 | 0.0190 | 0.0431 | 0.0210 | 66% |
| 32 | 0.0121 | 0.0372 | 0.0171 | 70% |
| 64 | 0.0083 | 0.0340 | 0.0339 | 72% |

"Uniform" replicates one utterance, so there is no padding: that is the batch's
true scaling curve and it keeps improving to 64. In arrival order, throughput
goes from 742 tok/s at batch 16 to 941 at batch 64, four times the batch and
twice the memory for 27% more. Over the whole librispeech-clean frozen set
(2620 cuts, median 5.79 s, max 34.95 s), batching in frozen-set order pays
2.35x the linear and 4.25x the attention work of the same audio at batch 16;
sorted by duration, 1.01x and 1.04x.

## What changed

- `attn_implementation` now reaches the backbones, and the wrapper admits it
  supports flash attention. The decoder takes the request unconditionally; the
  audio encoder takes it only if its own model class carries the kernel, since
  that is per encoder (in transformers 5.16.1 Whisper, Wav2Vec2, HuBERT and
  Qwen2-Audio do, Wav2Vec2-BERT and WavLM do not). A sub-config that already
  holds an explicit value keeps it, so `train.py`'s re-application still wins.
  Verified on the real checkpoint (mn5 job 45561844): batch 16 goes from 3
  distinct results in 5 repeats to 1.
- `run_inference.py` sorts by duration before batching.
- `docs/generation_eval.md` and the `--batch-size` help are corrected.

Anything that decodes should now pass it explicitly:

```python
MELTForCausalLM.from_pretrained(ckpt, dtype=torch.bfloat16,
                                attn_implementation="flash_attention_2")
```

Filed as MELT-proj/eval#7 for the eval harness, which needs the same two
changes.

## What does not change

**A batch will not reproduce `batch_size=1`, and that is not a bug to chase.**
Under `flash_attention_2` each configuration repeats exactly, but batch 16 still
differs from batch 1 on 6 of 16 rows, against 2 of 16 under sdpa. Different
shapes mean different reduction orders in bfloat16 and no attention
implementation changes that. A batched WER is comparable with another batched
WER at the same width, not with a single-sample number. `docs/generation_eval.md`
and the `--batch-size` help both claimed otherwise; they were wrong and are
fixed.

## The H100 slowdown does not reproduce

On MN5's H100, same checkpoint and code, batch 1 runs at 132 tokens/s, i.e.
7.6 ms per token, against the 1.35 s per token reported on the artemis H100.
That is roughly 180x, so the artemis figure is environment-specific rather than
a property of the modeling code, and worth chasing separately on that node.

**Comment by g8a9** [comment]
Follow-up, and a correction to the comment above.

**Correction.** I wrote that batch 16 differs from batch 1 on "6 of 16 rows
under `flash_attention_2`, against 2 of 16 under sdpa". That 6 of 16 is wrong.
It came from a probe that wrote `_attn_implementation` onto an already-built
model, including the audio encoder, which has no flash kernel — a configuration
nobody runs. Loading each configuration properly, flash attention is 2 of 16,
the same as sdpa. Flash attention is not worse for cross-width agreement.

**The real variable is the dtype, not the attention implementation.** Batch 16
against batch 1 over the same 16 utterances, alongside the first-step logit
perturbation the batch introduces and the top-two margin it has to cross to
flip a greedy choice (mn5 job 45563253):

| configuration | transcripts differing | median \|Δlogit\| | max \|Δlogit\| | median top-2 margin |
|---|---|---|---|---|
| bfloat16, sdpa as shipped | 2 of 16 | 1.6e-01 | 2.5e-01 | 4.16 |
| bfloat16, sdpa on deterministic backends | 4 of 16 | 1.8e-01 | 2.7e-01 | 4.13 |
| bfloat16, eager | 1 of 16 | 1.9e-01 | 4.3e-01 | 4.09 |
| bfloat16, flash_attention_2 | 2 of 16 | 1.6e-01 | 2.8e-01 | 4.13 |
| float32, eager | **0 of 16** | 1.6e-04 | 1.4e-03 | 4.07 |
| float32, sdpa on deterministic backends | **0 of 16** | 1.8e-04 | 1.0e-03 | 4.07 |

Every bfloat16 row perturbs the logits by about the same 0.2 whatever the
kernel; the implementation only shuffles which utterances land on the wrong
side of a close call, which is why the counts wander between 1 and 4 without
meaning anything. In each bfloat16 configuration exactly one row had a top-two
margin smaller than the perturbation, and in float32 none did.

float32 shrinks the perturbation by roughly a thousand and every transcript
then agrees. That is not a guarantee: the perturbation is smaller, not zero, so
a genuinely tied pair would still flip. It costs about twice the memory and
throughput, so it is a way to settle an argument about one specific number
rather than a campaign default. `torch.backends.cuda.matmul.allow_tf32` is
False in this image, so those rows really are float32 and not TF32.

Every configuration above repeats itself exactly, including the two that use
sdpa, because the deterministic-backend pin and flash attention both avoid the
cuDNN path. `docs/generation_eval.md` now carries this table.
